### ISSUE TITLE
FTBFS with openssl 1.1.0

### ISSUE THREAD

**Issue by ellert** [italiangrid/voms#50]
voms does not compile with openssl 1.1.0, as reported to Debian BTS:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=828595

**Comment by andreaceccanti** [comment]
Hello Mattias, 

this is a known issue, but we hope to fix with the next VOMS release (that will just provide Openssl 1.1.0 compatibility) and which is scheduled for mid-October (i.e. in a month from now).

**Comment by ellert** [comment]
I know you are aware of it. I just wanted to keep track of it and be able to mark the BTS bug as forwarded.

**Comment by ellert** [comment]
Hi!

Debian just increased the severity of the bug in BTS so it is now release critical. If not fixed before 2016-11-25, the package - and those depending on it - will be removed.

**Comment by andreaceccanti** [comment]
Hi Mattias,

we're working on it. I think we can make a release before the 2016-11-25 deadline.

**Comment by msalle** [comment]
Hi Andrea,
could let me know when you have a reasonably working development version, so that I can also runtime check LCMAPS and the like, which has the same deadline ?

**Comment by andreaceccanti** [comment]
I will.

A.

**Comment by giacomini** [comment]
Hi Mattias

I've started the porting, but I'm encountering a few problems with the new OpenSSL API that I'm not able to solve, even after looking into the source code. What's the best place to ask for some assistance? Should I put a comment in the original Debian bug? Can I contact directly the reporter of that bug (Kurt Roeckx)? something else?

**Comment by msalle** [comment]
Hi Francesco,
in my experience the best is to ask on the openssl-dev mailing list. What type of issues do you encounter? Most of the changes are now opaque struct members, for which there are mostly getters, plus the deprecation of the asn1_mac.h macros. Perhaps Mattias has better suggestions?

**Comment by giacomini** [comment]
Hi Mischa, the key of the problem is the word "mostly" :-) I can't find a way to express what is at https://github.com/italiangrid/voms/blob/master/src/socklib/Server.cpp#L362. I'm writing some notes on my attempts; I'll post them later.
The removal (not only deprecation) of the asn1_mac.h macros is another problem that I've not yet addressed. But in the worst case I can always copy those macros into the voms code base and remove them later.

**Comment by msalle** [comment]
Hi Francesco,
from looking around a bit in the openssl-1.1 code, e.g. https://github.com/openssl/openssl/blob/OpenSSL_1_1_0b/crypto/bio/bss_sock.c#L56-L65, I would think it's better to run BIO_new() with a custom BIO_METHOD (created using BIO_meth_new()), for which you can get/set the struct elements using BIO_meth_get_write()/BIO_meth_set_write() etc.
Not sure if this is the best way, but looks doable and reasonable.
I would certainly not try to reinsert the deprecated asn1_mac.h macros...
I updated lcmaps-plugins-verify-proxy but that needed only subset (https://ndpfsvn.nikhef.nl/viewvc/mwsec/trunk/lcmaps-plugins-verify-proxy/src/verify-proxy/verify-lib/src_internal/_verify_proxy_certinfo.c?annotate=18683 in particular after lines 87 and 143).
Globus has updates in https://github.com/globus/globus-toolkit/blame/globus_6_branch/gsi/proxy/proxy_ssl/source/library/proxycertinfo.c and that should probably have anything you need concerning the asn1 macro replacements.

**Comment by giacomini** [comment]
That's what I'm trying to do, but the API doesn't use `const` coherently, so, for example, the getter `BIO_meth_get_write(BIO_METHOD *biom)` takes a non-const BIO_METHOD; so I cannot use what I get back from `const BIO_METHOD *BIO_s_socket(void)`. Unless I use a cast, of course, which is what I'll do since I know from the code that the getter doesn't change it, but it's not clean. Apart from that, I'm not sure which type and name to use for the call to `BIO_METHOD *BIO_meth_new(int type, const char *name)`; I would like to take them from what `BIO_s_socket` gives me, but there are no getters available for that, they exist only for the BIO that uses that BIO_METHOD!
Thanks for the hints on the macros.

**Comment by giacomini** [comment]
That's what I'm trying to do, but the API doesn't use `const` coherently. For example `BIO_meth_get_write(BIO_METHOD *biom))` takes a non-const parameter, so I cannot pass what `const BIO_METHOD *BIO_s_socket(void);` returns. I can (and will) use a cast in this case, because I know from the source code that the argument is not modified, but it's not clean.

Then, I'm not sure what to pass as type and name to `BIO_METHOD *BIO_meth_new(int type, const char *name)`. I would like to take them from the BIO_METHOD returned by `BIO_s_socket`, but there are no getters for those; curiously they exist for a BIO, but I would prefer not to build a temporary BIO just to get those.

Thanks for the hints on the macros.

**Comment by msalle** [comment]
If I understand correctly, ``BIO_s_socket()`` returns the default  ``BIO_METHOD`` which is rightfully ``const``, but you want to make your own, since you want to adjust the ``bwrite``/``bread`` functions, so you need to call ``BIO_meth_new`` with the next free type, which you can get using ``BIO_get_new_index()`` plus a new name, e.g. "Globus I/O" or something like that ?

**Comment by giacomini** [comment]
That's also what I'm doing for the type in fact, but I'm not sure that's correct. Thanks for confirming that that's the right way.

About `const`, I agree that `BIO_s_socket()` does the right thing; it's the various getters that seem wrong (from what I understand).

Anyway, from this discussion I think I can move forward. Thanks.

**Comment by msalle** [comment]
Note that I'm not a real expert, it's also for me just looking at code and manpages...
What do you think is wrong with the various getters (and which ones actually?).
You can always ask on the openssl-dev mailing list...

**Comment by giacomini** [comment]
This is what I want to do
```
#include <openssl/bio.h>

int main()
{
  int const type = BIO_get_new_index();
  BIO_METHOD* voms_biom = BIO_meth_new(type, "VOMS I/O");
  BIO_METHOD const* sock_biom = BIO_s_socket();
  BIO_meth_set_write(voms_biom, BIO_meth_get_write(sock_biom));
  /* BIO_meth_set_write(voms_biom, BIO_meth_get_write((BIO_METHOD*)sock_biom)); */
}
```
But it doesn't compile. It does if you apply a cast to `sock_biom` to remove the `const`. Not nice.

**Comment by ellert** [comment]
I agree that the signatures of the getters should use const. The getters do not modify the object.

**Comment by msalle** [comment]
Ah, I think I agree: the ``BIO_meth_get_write()`` should have a ``const BIO_METHOD *`` argument, since it does not change it. Likewise obviously for the ``BIO_meth_get_read()``.
That's probably something to report to openssl (as issue via github).

**Comment by msalle** [comment]
We agree (-; For the time being an explicit cast should be harmless.

**Comment by kouril** [comment]
I've just stumbled on this progress and ticket. If it helps for the sslutils part, take a look at canl-c, the openssl-1.1 branch where we pushed the necessary changes earlier this week.

**Comment by giacomini** [comment]
Thanks, Daniel. I've already done sslutils part, but I'll compare with what you have done to see if I've made any mistake. Unfortunately the missing part concerns the contents of proxycertinfo.c, which you have in your repo but apparently don't use so you didn't need to do the porting.

**Comment by msalle** [comment]
Hi Francesco,
I would certainly also look at the Globus version (see my previous comment above) which contains also a proxycertinfo.c, although the VOMS and Globus versions have diverged quite a bit over the years, you can still get a reasonable idea how one can do things.

**Comment by ellert** [comment]
Globus's reimplementation of proxycertinfo and proxypolicy for openssl 1.1 is here:
https://github.com/globus/globus-toolkit/tree/globus_6_branch/gsi/proxy/proxy_ssl/source/library

Globus now also uses the proxycertinfo struct and the associated functions that are defined in the openssl library in many places. The definition in the globus-proxy-ssl is marked as deprecated, and is mainly used for the implementation of non-RFC proxies.

**Comment by giacomini** [comment]
Thanks Mattias, I found the globus code. I'll try to integrate it as much as I can. After that, the next hurdle is the encoding/decoding of the AC, which currently uses the removed macros. I'm reimplementing it in the "right" way. If only I could find some documentation on it, instead of looking into the code...

**Comment by ellert** [comment]
Hi!

How are things progressing? Can I help in any way?

**Comment by giacomini** [comment]
Hi Mattias.

There is some progress. The code using the removed macros has been ported to OpensSSL 1.1 for the encoding/decoding of the proxy information, but not yet fully for the AC. But porting is just the first step, so there is still quite some work to do.

As soon as there is a stable situation I'll let you know, so that you can at least try the build.

**Comment by ellert** [comment]
In order not to be removed from Debian I built voms using the ssl11 branch in Debian unstable. This compiled, so I could close the FTBFS bug. But it is not fully functional, so It should be updated again when it works. The full freeze for debian 9 is 2017-02-05, and migration from unstable to testing takes 10 days, so the absolutely last day to upload the version that will be in Debian 9 is 2017-01-25. And preferably it should not be done at the very last minute...

**Comment by andreaceccanti** [comment]
I wonder what is preventing Debian to have an openssl-1.0 compatibility package, like Fedora does AFAIR. I am quite sure we won't make the deadline, as more work is still to be done for a properly tested transition to 1.1...

**Comment by ellert** [comment]
There is an openssl 1.0 compatibility library in debian too. However, since there are a lot of software (lcgdm, myproxy, fts, ...) that uses both the globus libraries and the voms libraries it would be very problematic to have the globus libraries compiled against openssl 1.1 and the voms libraries compiled against openssl 1.0 and then compile against both the globus and voms libraries since you would then have both openssl library versions in the same process...

**Comment by andreaceccanti** [comment]
So the idea is that we have a problem on fedora as well...

**Comment by ellert** [comment]
Yes, but the deadlines are not as close in time. (alpha freeze 2017-02-28, beta freeze 2017-04-25, final freeze 2017-05-23).

**Comment by msalle** [comment]
Even if VOMS manages, dependencies will probably not be able to, in such short notice...
I do fully agree with Mattias that building VOMS against 1.0 while globus is built against 1.1 will be a nightmare (read: probably one big segfault).

**Comment by giacomini** [comment]
Sure. We'll do our best to meet the deadline.

**Comment by andreaceccanti** [comment]
We will provide you a tag tonight that builds and where the API and server works as expected. No effort will be invested in checking that the old clients work as expected, even if I do not expect big regressions there (and anyway, these are not officially supported by us).

**Comment by andreaceccanti** [comment]
The tag is https://github.com/italiangrid/voms/releases/tag/v2.1.0-rc0

**Comment by ellert** [comment]
Many thanks.

I uploaded the new version to Debian unstable last night:
https://packages.qa.debian.org/v/voms/news/20170125T213541Z.html
With 10 days in unstable before migrating to testing this should just about make it before the freeze on February 5.

I also built it for Fedora Rawhide this morning:
https://koji.fedoraproject.org/koji/buildinfo?buildID=836347

**Comment by andreaceccanti** [comment]
Thank you Mattias.
@msalle, let us know if VOMS 2.1.0 breaks lcmaps...

**Comment by msalle** [comment]
Hi,
It seems to work fine!

I found one minor issue
https://github.com/italiangrid/voms/blob/develop-2.1.x/src/sslutils/proxycertinfo.c#L350
This assertion goes wrong if you're also using globus (as we are doing in the lcas-lcmaps-gt-interface) since globus already creates the NID. Adding something like
```
  if (OBJ_txt2nid(PROXYCERTINFO_OLD_OID)!=NID_undef)	{
      init_done = 1;
      return;
  }
```
just before the `OBJ_create` solves the problem.
I also checked the clients and they actually seem to work too. It only produces a harmless error, which comes from https://github.com/italiangrid/voms/blob/develop-2.1.x/src/sslutils/sslutils.c#L480 (see also line before), which is not it self part of the clients (before you say that the clients are no longer supported (-: )

**Comment by giacomini** [comment]
Thanks for the good news!
We'll go through all the asserts. I often put them in the code as a placeholder for proper error handling and indeed some may need to be reviewed.

Mattias, can you remind me what the procedure is now for modifications?

**Comment by msalle** [comment]
Hi,
I'm afraid I found one more issue: https://github.com/italiangrid/voms/blob/develop-2.1.x/src/api/ccapi/voms_apic.h#L41
gssapi.h defines it with a _starting_ and a _trailing_ underscore. See e.g. https://github.com/globus/globus-toolkit/blob/globus_6_branch/gsi/gssapi/source/library/gssapi.h and https://github.com/krb5/krb5/blob/krb5-1.15/src/lib/gssapi/generic/gssapi.hin

**Comment by giacomini** [comment]
But do you see a specific build- or run-time problem? In fact the globus gssapi.h defines both forms (https://github.com/globus/globus-toolkit/blob/globus_6_branch/gsi/gssapi/source/library/gssapi.h#L18). BTW, technically speaking identifiers starting with an underscore followed by a capital letter are reserved and shouldn't be used (at least in C++).

**Comment by msalle** [comment]
Yes, I've seen it fail: if I build against my system gssapi.h, i.e. not the globus one (this was an LCMAPS plugin which does not really use Globus itself). That gssapi.h only defines `_GSSAPI_H_`, i.e. with a starting **and** ending underscore, conform https://github.com/globus/globus-toolkit/blob/globus_6_branch/gsi/gssapi/source/library/gssapi.h#L20-L24
But since https://github.com/italiangrid/voms/blob/develop-2.1.x/src/api/ccapi/voms_apic.h#L34-L41 only checks for **either** starting **or** ending, it fails.

**Comment by andreaceccanti** [comment]
@msalle , no changes were introduced with this version that affect that header, so I would say this issue is not 2.1.0 specific. Can you confirm whether you observe the build failure also with 2.0.14, for instance? Thanks

**Comment by msalle** [comment]
Hi,
You're right, I agree it's indeed an older issue and indeed also fails with 2.0.14.
For a simple test program:
```
#include <gssapi.h>
#include <voms/voms_apic.h>

int main(void)	{ return 0; }
```
Compiling against krb-based gssapi.h fails, compiling against globus-based gssapi.h succeeds.

**Comment by andreaceccanti** [comment]
Ok, thanks @msalle. I've created a separate issue to track this problem (#54)

**Comment by ellert** [comment]
My observation so far.

The command line client seems to work. When using voms-proxy-init the following lines are printed to the screen:

oid exists
Function: OBJ_create

but execution continues, and the proxy that is created seems to be correct.

The default proxy version in the absence of any command line options is still the legacy proxy though, which is different form the java client that makes RFC proxies by default.

The biggest issue I have found so far is with myproxy. When myproxy is compiled with voms support using the new library version the test suite fails. Voms support in myproxy is implemented using a plugin loaded at runtime which uses the voms library. If I move the plugin out of the way so that it can not be loaded and rerun the test suite it succeeds. This might be caused by the same issue previously mentioned by @msalle regarding using globus and voms libraries together.

**Comment by ellert** [comment]
Using the change to src/sslutils/proxycertinfo.c proposed by @msalle fixes the issue with the myproxy test suite. I created pull request #55 with this change.

**Comment by giacomini** [comment]
Hi. I propose a larger change; see issue #60 and pull request #61.

**Comment by giacomini** [comment]
I have also create pull request #62 to fix issue #54. Could you please have a look?

**Comment by msalle** [comment]
It looks fine to me, thanks!

**Comment by ellert** [comment]
Last Wednesday 1 February, I built an updated voms package for Fedora Rawhide. This package corresponds to the 2.1.0rc0 version + the pull requests #55, #56, #57 and #58. I then built the latest myproxy version in Fedora Rawhide using this voms version and the test suite was successful (didn't trigger an assertion in the voms library).

Early yesterday morning, the voms package with the 2.1.0rc0 version migrated to Debian testing after 10 days in Debian unstable. Late last night the new Debian release was declared to be in Final Freeze, and all further migrations to testing must have an explicit unblock approval from the Debian Release Team.

Yesterday I uploaded an updated package to Debian containing the same patches as the Fedora update mentioned above. I then requested an unblock approval for this update which was granted, so it looks like there will be a working voms library in the next Debian Release.
